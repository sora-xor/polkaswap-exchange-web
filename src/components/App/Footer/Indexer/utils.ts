import { IndexerType } from '@soramitsu/soraneo-wallet-web/lib/consts';

export function getIndexerName(type: IndexerType): string {
  return type.charAt(0).toUpperCase() + type.slice(1).toLowerCase();
}
