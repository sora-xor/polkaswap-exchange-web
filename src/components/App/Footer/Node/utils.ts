import { countryCodeEmoji } from 'country-code-emoji';

import store from '@/store';

export function formatLocation(code: string): Nullable<{ flag: string; name?: string }> {
  try {
    const isoCode = code.toUpperCase();
    const flag = countryCodeEmoji(isoCode);
    const location = { flag, name: '' };
    const displayRegions = store.state.settings.displayRegions;
    if (!displayRegions) {
      return location;
    }
    const name = displayRegions.of(isoCode);
    if (name) {
      location.name = name;
    }
    return location;
  } catch (error) {
    console.warn('Unsupported format of location in env.json', error);
    return null;
  }
}
