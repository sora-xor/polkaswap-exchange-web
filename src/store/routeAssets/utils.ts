import { FPNumber, Operation } from '@sora-substrate/util/build';
import { Asset, AccountAsset } from '@sora-substrate/util/build/assets/types';
import { api } from '@soramitsu/soraneo-wallet-web';
import { FiatPriceObject } from '@soramitsu/soraneo-wallet-web/lib/services/subquery/types';

import { Recipient } from './types';

import type { WhitelistArrayItem } from '@sora-substrate/util/build/assets/types';

export default {
  validate(recipient: Recipient) {
    return (
      this.wallet(recipient.wallet) &&
      this.asset(recipient.asset) &&
      this.usd(recipient.usd) &&
      this.amount(recipient.amount)
    );
  },

  name(name: Nullable<string>) {
    return name;
  },

  wallet(wallet: string) {
    return api.validateAddress(wallet);
  },

  asset(asset: Asset | AccountAsset) {
    return asset;
  },

  usd(usd: Nullable<number>) {
    return usd && !isNaN(usd);
  },

  amount(amount: Nullable<number>) {
    return amount && !isNaN(amount) && Number.isFinite(amount);
  },
};

export function getTokenEquivalent(
  priceObject: FiatPriceObject,
  asset: Asset | AccountAsset | WhitelistArrayItem,
  usd: number | string
): FPNumber {
  return new FPNumber(usd).div(FPNumber.fromCodecValue(priceObject[asset.address], asset.decimals));
}

export function getAssetUSDPrice(asset, fiatPriceObject) {
  if (!asset) return null;
  return FPNumber.fromCodecValue(fiatPriceObject[asset.address], asset.decimals);
}
