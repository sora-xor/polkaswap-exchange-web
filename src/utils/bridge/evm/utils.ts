import { Operation } from '@sora-substrate/sdk';

import { evmBridgeApi } from '@/utils/bridge/evm/api';

import type { EvmHistory } from '@sora-substrate/sdk/build/bridgeProxy/evm/types';

export const isOutgoingTx = (tx: EvmHistory): boolean => {
  return tx.type === Operation.EvmOutgoing;
};

export const isUnsignedTx = (tx: EvmHistory): boolean => {
  if (isOutgoingTx(tx)) {
    return !tx.blockId && !tx.txId;
  } else {
    return !tx.externalHash;
  }
};

export const getTransaction = (id: string): EvmHistory => {
  const tx = evmBridgeApi.getHistory(id) as EvmHistory;

  if (!tx) throw new Error(`[Bridge]: Transaction is not exists: ${id}`);

  return tx;
};

export const updateTransaction = (id: string, params = {}): void => {
  const tx = getTransaction(id);
  evmBridgeApi.saveHistory({ ...tx, ...params });
};
