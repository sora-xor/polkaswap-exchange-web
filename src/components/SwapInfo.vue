<template>
  <div class="swap-info-container">
    <template v-if="showPrice || showSlippageTolerance">
      <div v-if="showPrice" class="swap-info">
        <span>{{ t('exchange.price') }}</span>
        <span class="swap-info-value">{{ priceValue }}</span>
        <s-button class="el-button--switch-price" type="action" size="small" icon="swap" @click="handleSwitchPrice" />
      </div>
      <div v-if="showSlippageTolerance && connected" class="swap-info swap-info--slippage-tolerance">
        <span>{{ t('swap.slippageTolerance') }}</span>
        <s-button class="swap-info-value" type="link" @click="openSettingsDialog">
          {{ slippageTolerance }}%
        </s-button>
      </div>
      <settings v-if="showSlippageTolerance" :visible.sync="showSettings" />
    </template>
    <template v-else>
      <div class="swap-info swap-info--min-received">
        <s-tooltip v-if="showTooltips" class="swap-info-icon" popper-class="info-tooltip info-tooltip--swap" border-radius="mini" :content="t('swap.minReceivedTooltip')" theme="light" placement="right-start" animation="none" :show-arrow="false">
          <s-icon name="info" size="16" />
        </s-tooltip>
        <span>{{ t(`swap.${isExchangeB ? 'maxSold' : 'minReceived'}`) }}</span>
        <span class="swap-info-value">{{ formattedMinMaxReceived }}<span class="asset-title">{{ getAssetSymbolText }}</span></span>
      </div>
      <!-- <div class="swap-info">
        <s-tooltip v-if="showTooltips" class="swap-info-icon" popper-class="info-tooltip info-tooltip--swap" border-radius="mini" :content="t('swap.priceImpactTooltip')" theme="light" placement="right-start" animation="none" :show-arrow="false">
          <s-icon name="info" size="16" />
        </s-tooltip>
        <span>{{ t('swap.priceImpact') }}</span>
        <span :class="'swap-info-value ' + priceImpactClass">{{ priceImpact }}%</span>
      </div> -->
      <div class="swap-info">
        <s-tooltip v-if="showTooltips" class="swap-info-icon" popper-class="info-tooltip info-tooltip--swap" border-radius="mini" :content="t('swap.liquidityProviderFeeTooltip', { liquidityProviderFee: 0.3})" theme="light" placement="right-start" animation="none" :show-arrow="false">
          <s-icon name="info" size="16" />
        </s-tooltip>
        <span>{{ t('swap.liquidityProviderFee') }}</span>
        <span class="swap-info-value">{{ formattedLiquidityProviderFee }}<span class="asset-title">{{ xorSymbol }}</span></span>
      </div>
      <!-- TODO 4 alexnatalia: Show if logged in and have info about Network Fee -->
      <div v-if="connected" class="swap-info">
        <s-tooltip v-if="showTooltips" class="swap-info-icon" popper-class="info-tooltip info-tooltip--swap" border-radius="mini" :content="t('swap.networkFeeTooltip')" theme="light" placement="right-start" animation="none" :show-arrow="false">
          <s-icon name="info" size="16" />
        </s-tooltip>
        <span>{{ t('swap.networkFee') }}</span>
        <span class="swap-info-value">{{ formattedNetworkFee }}<span class="asset-title">{{ xorSymbol }}</span></span>
      </div>
    </template>
  </div>
</template>

<script lang="ts">
import { Component, Mixins, Prop } from 'vue-property-decorator'
import { Action, Getter } from 'vuex-class'
import { KnownSymbols, CodecString, AccountAsset } from '@sora-substrate/util'

import TranslationMixin from '@/components/mixins/TranslationMixin'
import NumberFormatterMixin from '@/components/mixins/NumberFormatterMixin'
import { isWalletConnected } from '@/utils'
import { lazyComponent } from '@/router'
import { Components } from '@/consts'

const namespace = 'swap'

@Component({
  components: {
    Settings: lazyComponent(Components.Settings)
  }
})
export default class SwapInfo extends Mixins(TranslationMixin, NumberFormatterMixin) {
  @Getter('tokenFrom', { namespace }) tokenFrom!: AccountAsset
  @Getter('tokenTo', { namespace }) tokenTo!: AccountAsset
  @Getter('isTokenFromPrice', { namespace }) isTokenFromPrice!: boolean
  @Getter('minMaxReceived', { namespace }) minMaxReceived!: CodecString
  @Getter('isExchangeB', { namespace }) isExchangeB!: boolean
  @Getter('liquidityProviderFee', { namespace }) liquidityProviderFee!: CodecString
  @Getter('networkFee', { namespace }) networkFee!: CodecString
  @Action('setTokenFromPrice', { namespace }) setTokenFromPrice!: (isTokenFromPrice: boolean) => Promise<void>

  @Getter('price', { namespace: 'prices' }) price!: string
  @Getter('priceReversed', { namespace: 'prices' }) priceReversed!: string

  @Getter slippageTolerance!: number

  @Prop({ default: false, type: Boolean }) readonly showPrice!: boolean
  @Prop({ default: true, type: Boolean }) readonly showTooltips!: boolean
  @Prop({ default: false, type: Boolean }) readonly showSlippageTolerance!: boolean

  showSettings = false

  get connected (): boolean {
    return isWalletConnected()
  }

  get priceValue (): string {
    const fromSymbol = this.tokenFrom ? this.tokenFrom.symbol : ''
    const toSymbol = this.tokenTo ? this.tokenTo.symbol : ''
    if (this.isTokenFromPrice) {
      return `${this.price} ${fromSymbol} / ${toSymbol}`
    }
    return `${this.priceReversed} ${toSymbol} / ${fromSymbol}`
  }

  get formattedNetworkFee (): string {
    return this.formatCodecNumber(this.networkFee)
  }

  get formattedLiquidityProviderFee (): string {
    return this.formatCodecNumber(this.liquidityProviderFee)
  }

  get formattedMinMaxReceived (): string {
    const decimals = (this.isExchangeB ? this.tokenFrom : this.tokenTo)?.decimals
    return this.formatCodecNumber(this.minMaxReceived, decimals)
  }

  // TODO: [Release 2]
  // get priceImpact (): string {
  //   return '0'
  // }
  // get priceImpactClass (): string {
  //   if (+this.priceImpact > 0) {
  //     return 'price-impact-positive'
  //   }
  //   if (+this.priceImpact < 0) {
  //     return 'price-impact-negative'
  //   }
  //   return ''
  // }

  get xorSymbol (): string {
    return ' ' + KnownSymbols.XOR
  }

  get getAssetSymbolText (): string {
    if (this.isExchangeB) {
      return this.tokenFrom ? ` ${this.tokenFrom.symbol}` : ''
    }
    return this.tokenTo ? ` ${this.tokenTo.symbol}` : ''
  }

  handleSwitchPrice (): void {
    this.setTokenFromPrice(!this.isTokenFromPrice)
  }

  openSettingsDialog (): void {
    this.showSettings = true
  }
}
</script>

<style lang="scss">
.info-tooltip--swap {
  margin-left: #{$inner-spacing-mini / 2} !important;
}
.el-button--switch-price {
  @include switch-button-inherit-styles;
  &.s-action.s-small i {
    margin-top: 0;
    margin-left: 0;
  }
}
</style>

<style lang="scss" scoped>
@include info-line;
.swap-info {
  &--slippage-tolerance,
  &--min-received {
    margin-top: $inner-spacing-small;
  }
  .price-impact {
    &-positive {
      color: var(--s-color-status-success);
    }
    &-negative {
      color: var(--s-color-status-error);
    }
  }
  &-value.el-button {
    margin-right: 0;
    height: var(--s-font-size-small);
    padding: 0;
    color: inherit;
  }
  .el-button--switch-price {
    margin-right: 0;
    margin-left: $inner-spacing-mini;
    @include switch-button;
  }
}
</style>
