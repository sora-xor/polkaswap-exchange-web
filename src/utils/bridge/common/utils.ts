import { Operation, isEthOperation, isEvmOperation, isSubstrateOperation } from '@sora-substrate/util';
import { api as soraApi } from '@soramitsu/soraneo-wallet-web';
import { ethers } from 'ethers';

import { isUnsignedTx as isUnsignedEthTx } from '@/utils/bridge/eth/utils';
import { isUnsignedTx as isUnsignedEvmTx } from '@/utils/bridge/evm/utils';
import { isUnsignedTx as isUnsignedSubTx } from '@/utils/bridge/sub/utils';
import ethersUtil from '@/utils/ethers-util';

import type { ApiPromise } from '@polkadot/api';
import type { IBridgeTransaction } from '@sora-substrate/util';
import type { EthHistory } from '@sora-substrate/util/build/bridgeProxy/eth/types';
import type { EvmHistory } from '@sora-substrate/util/build/bridgeProxy/evm/types';
import type { SubHistory } from '@sora-substrate/util/build/bridgeProxy/sub/types';

export const waitForEvmTransactionMined = async (
  tx: ethers.TransactionResponse | null,
  replaceCallback?: (tx: ethers.TransactionResponse | null) => void
): Promise<ethers.TransactionReceipt | null> => {
  if (!tx) throw new Error('[waitForEvmTransactionMined]: tx cannot be empty!');

  try {
    const startBlock = tx.blockNumber ?? (await ethersUtil.getBlockNumber());
    const replaceableTx = tx.replaceableTransaction(startBlock);
    const txReceipt = await replaceableTx.wait();

    return txReceipt;
  } catch (error) {
    if (ethers.isError(error, 'TRANSACTION_REPLACED')) {
      const replacedTx = error.replacement;

      replaceCallback?.(replacedTx);

      return await waitForEvmTransactionMined(replacedTx, replaceCallback);
    }

    throw error;
  }
};

export const getEvmTransactionRecieptByHash = async (
  transactionHash: string
): Promise<{ fee: string; blockHash: string; blockNumber: number; from: string } | null> => {
  try {
    const receipt = await ethersUtil.getEvmTransactionReceipt(transactionHash);

    if (!receipt) throw new Error(`Transaction receipt "${transactionHash}" not found`);

    const { fee, from, blockNumber, blockHash } = receipt;

    return { fee: fee.toString(), blockHash, blockNumber, from };
  } catch (error) {
    return null;
  }
};

export const getBlockEventsByTxIndex = async (blockHash: string, index: number, api: ApiPromise) => {
  const blockEvents = await soraApi.system.getBlockEvents(blockHash, api);
  const transactionEvents = blockEvents.filter(
    ({ phase }) => phase.isApplyExtrinsic && phase.asApplyExtrinsic.toNumber() === index
  );

  return transactionEvents;
};

export const getTransactionEvents = async (blockHash: string, transactionHash: string, api: ApiPromise) => {
  const extrinsics = await soraApi.system.getExtrinsicsFromBlock(blockHash, api);
  const extrinsicIndex = extrinsics.findIndex((ext) => ext.hash.toString() === transactionHash);

  if (extrinsicIndex === -1) throw new Error(`Unable to find extrinsic "${transactionHash}" in block "${blockHash}"`);

  const transactionEvents = await getBlockEventsByTxIndex(blockHash, extrinsicIndex, api);

  return transactionEvents;
};

export const isOutgoingTransaction = (transaction: Nullable<IBridgeTransaction>): boolean => {
  if (!transaction?.type) return false;

  return [Operation.EthBridgeOutgoing, Operation.EvmOutgoing, Operation.SubstrateOutgoing].includes(transaction.type);
};

export const isUnsignedTx = (transaction: Nullable<IBridgeTransaction>): boolean => {
  if (!transaction?.type) return true;

  if (isEthOperation(transaction.type)) return isUnsignedEthTx(transaction as EthHistory);
  if (isEvmOperation(transaction.type)) return isUnsignedEvmTx(transaction as EvmHistory);
  if (isSubstrateOperation(transaction.type)) return isUnsignedSubTx(transaction as SubHistory);

  return true;
};
