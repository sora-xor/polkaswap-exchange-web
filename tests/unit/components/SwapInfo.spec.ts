import { VueConstructor } from 'vue';
import Vuex from 'vuex';
import { shallowMount } from '@vue/test-utils';

import SwapInfo from '@/components/SwapInfo.vue';

import { SoramitsuElementsImport, localVue, TranslationMock } from '../../utils';
import { MOCK_SWAP_INFO, MOCK_FIAT_PRICE_AND_APY_OBJECT } from '../../mocks/components/SwapInfo';

const NumberFormatterMixin = (vue: VueConstructor) => vue.mixin({ name: 'NumberFormatterMixin' });

const FormattedAmountMixin = (vue: VueConstructor) => vue.mixin({ name: 'FormattedAmountMixin' });

export const useDescribe = (name: string, component: VueConstructor<Vue>, fn: jest.EmptyFunction) =>
  describe(name, () => {
    beforeAll(() => {
      SoramitsuElementsImport(localVue);
      TranslationMock(component);
      NumberFormatterMixin(component);
      FormattedAmountMixin(component);
    });
    fn();
  });

useDescribe('SwapInfo.vue', SwapInfo, () => {
  MOCK_SWAP_INFO.map((item) =>
    it(`[${item.title}]: should be rendered correctly`, () => {
      const store = new Vuex.Store({
        modules: {
          swap: {
            namespaced: true,
            state: {
              liquidityProviderFee: item.liquidityProviderFee,
              isExchangeB: item.isExchangeB,
              rewards: item.rewards,
            },
            getters: {
              tokenFrom: () => item.tokenFrom,
              tokenTo: () => item.tokenTo,
              minMaxReceived: () => item.minMaxReceived,
              priceImpact: () => item.priceImpact,
              priceReversed: () => item.priceReversed,
              price: () => item.price,
            },
          },
        },
        getters: {
          fiatPriceAndApyObject: () => MOCK_FIAT_PRICE_AND_APY_OBJECT,
          isLoggedIn: () => item.isLoggedIn,
          networkFees: () => item.networkFees,
        },
      });
      const wrapper = shallowMount(SwapInfo, {
        localVue,
        store,
      });
      expect(wrapper.element).toMatchSnapshot();
    })
  );
});
