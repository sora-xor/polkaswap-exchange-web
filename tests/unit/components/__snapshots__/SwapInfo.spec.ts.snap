// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwapInfo.vue [With All Set Parameters]: should be rendered correctly 1`] = `
<div
  class="swap-info-container"
>
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="{first} per {second}"
    labeltooltip=""
    value="0.4"
  />
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="{first} per {second}"
    labeltooltip=""
    value="2.2"
  />
   
  <info-line-stub
    assetsymbol="VAL"
    isformatted="true"
    label="Minimum Received"
    labeltooltip="Your transaction will revert if there is a large, unfavorable price movement before it is confirmed."
    value="1.2345678"
  />
    
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="Price Impact"
    labeltooltip="The difference between the market price and estimated price due to trade size."
    value=""
  >
    <!---->
  </info-line-stub>
   
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="Route"
    labeltooltip=""
    value=""
  />
   
  <info-line-stub
    assetsymbol=" XOR"
    fiatvalue=""
    isformatted="true"
    label="Liquidity Provider Fee"
    labeltooltip="A portion of each trade ({liquidityProviderFee}%) goes to liquidity providers as a protocol incentive."
    value="0.003"
  />
   
  <info-line-stub
    assetsymbol=" XOR"
    fiatvalue="0.000861"
    isformatted="true"
    label="Network Fee"
    labeltooltip="Network fee is used to ensure SORA system's growth and stable performance."
    value="0.0007"
  />
</div>
`;

exports[`SwapInfo.vue [With Empty liquidityProviderFee]: should be rendered correctly 1`] = `
<div
  class="swap-info-container"
>
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="{first} per {second}"
    labeltooltip=""
    value="0.4"
  />
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="{first} per {second}"
    labeltooltip=""
    value="2.2"
  />
   
  <info-line-stub
    assetsymbol="VAL"
    isformatted="true"
    label="Minimum Received"
    labeltooltip="Your transaction will revert if there is a large, unfavorable price movement before it is confirmed."
    value="1.2345678"
  />
    
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="Price Impact"
    labeltooltip="The difference between the market price and estimated price due to trade size."
    value=""
  >
    <div
      class="value-status-wrapper"
    >
      <formatted-amount-stub
        assetsymbol=""
        class="swap-value"
        fontsizerate="normal"
        fontweightrate="normal"
        value="-0.1"
      >
        %
      </formatted-amount-stub>
    </div>
  </info-line-stub>
   
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="Route"
    labeltooltip=""
    value=""
  />
   
  <info-line-stub
    assetsymbol=" XOR"
    fiatvalue=""
    isformatted="true"
    label="Liquidity Provider Fee"
    labeltooltip="A portion of each trade ({liquidityProviderFee}%) goes to liquidity providers as a protocol incentive."
    value="NaN"
  />
   
  <info-line-stub
    assetsymbol=" XOR"
    fiatvalue="0.000861"
    isformatted="true"
    label="Network Fee"
    labeltooltip="Network fee is used to ensure SORA system's growth and stable performance."
    value="0.0007"
  />
</div>
`;

exports[`SwapInfo.vue [With Empty minMaxReceived]: should be rendered correctly 1`] = `
<div
  class="swap-info-container"
>
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="{first} per {second}"
    labeltooltip=""
    value="0.4"
  />
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="{first} per {second}"
    labeltooltip=""
    value="2.2"
  />
   
  <info-line-stub
    assetsymbol="VAL"
    isformatted="true"
    label="Minimum Received"
    labeltooltip="Your transaction will revert if there is a large, unfavorable price movement before it is confirmed."
    value="NaN"
  />
    
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="Price Impact"
    labeltooltip="The difference between the market price and estimated price due to trade size."
    value=""
  >
    <div
      class="value-status-wrapper"
    >
      <formatted-amount-stub
        assetsymbol=""
        class="swap-value"
        fontsizerate="normal"
        fontweightrate="normal"
        value="-0.1"
      >
        %
      </formatted-amount-stub>
    </div>
  </info-line-stub>
   
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="Route"
    labeltooltip=""
    value=""
  />
   
  <info-line-stub
    assetsymbol=" XOR"
    fiatvalue=""
    isformatted="true"
    label="Liquidity Provider Fee"
    labeltooltip="A portion of each trade ({liquidityProviderFee}%) goes to liquidity providers as a protocol incentive."
    value="0.003"
  />
   
  <info-line-stub
    assetsymbol=" XOR"
    fiatvalue="0.000861"
    isformatted="true"
    label="Network Fee"
    labeltooltip="Network fee is used to ensure SORA system's growth and stable performance."
    value="0.0007"
  />
</div>
`;

exports[`SwapInfo.vue [With Empty networkFee]: should be rendered correctly 1`] = `
<div
  class="swap-info-container"
>
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="{first} per {second}"
    labeltooltip=""
    value="0.4"
  />
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="{first} per {second}"
    labeltooltip=""
    value="2.2"
  />
   
  <info-line-stub
    assetsymbol="VAL"
    isformatted="true"
    label="Minimum Received"
    labeltooltip="Your transaction will revert if there is a large, unfavorable price movement before it is confirmed."
    value="1.2345678"
  />
    
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="Price Impact"
    labeltooltip="The difference between the market price and estimated price due to trade size."
    value=""
  >
    <div
      class="value-status-wrapper"
    >
      <formatted-amount-stub
        assetsymbol=""
        class="swap-value"
        fontsizerate="normal"
        fontweightrate="normal"
        value="-0.1"
      >
        %
      </formatted-amount-stub>
    </div>
  </info-line-stub>
   
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="Route"
    labeltooltip=""
    value=""
  />
   
  <info-line-stub
    assetsymbol=" XOR"
    fiatvalue=""
    isformatted="true"
    label="Liquidity Provider Fee"
    labeltooltip="A portion of each trade ({liquidityProviderFee}%) goes to liquidity providers as a protocol incentive."
    value="0.003"
  />
   
  <info-line-stub
    assetsymbol=" XOR"
    fiatvalue="0"
    isformatted="true"
    label="Network Fee"
    labeltooltip="Network fee is used to ensure SORA system's growth and stable performance."
    value="NaN"
  />
</div>
`;

exports[`SwapInfo.vue [With Error priceImpact]: should be rendered correctly 1`] = `
<div
  class="swap-info-container"
>
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="{first} per {second}"
    labeltooltip=""
    value="0.4"
  />
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="{first} per {second}"
    labeltooltip=""
    value="2.2"
  />
   
  <info-line-stub
    assetsymbol="VAL"
    isformatted="true"
    label="Minimum Received"
    labeltooltip="Your transaction will revert if there is a large, unfavorable price movement before it is confirmed."
    value="1.2345678"
  />
    
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="Price Impact"
    labeltooltip="The difference between the market price and estimated price due to trade size."
    value=""
  >
    <div
      class="value-status-wrapper error"
    >
      <formatted-amount-stub
        assetsymbol=""
        class="swap-value"
        fontsizerate="normal"
        fontweightrate="normal"
        value="-10"
      >
        %
      </formatted-amount-stub>
    </div>
  </info-line-stub>
   
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="Route"
    labeltooltip=""
    value=""
  />
   
  <info-line-stub
    assetsymbol=" XOR"
    fiatvalue=""
    isformatted="true"
    label="Liquidity Provider Fee"
    labeltooltip="A portion of each trade ({liquidityProviderFee}%) goes to liquidity providers as a protocol incentive."
    value="0.003"
  />
   
  <info-line-stub
    assetsymbol=" XOR"
    fiatvalue="0.000861"
    isformatted="true"
    label="Network Fee"
    labeltooltip="Network fee is used to ensure SORA system's growth and stable performance."
    value="0.0007"
  />
</div>
`;

exports[`SwapInfo.vue [With Logged Out]: should be rendered correctly 1`] = `
<div
  class="swap-info-container"
>
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="{first} per {second}"
    labeltooltip=""
    value="0.4"
  />
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="{first} per {second}"
    labeltooltip=""
    value="2.2"
  />
   
  <info-line-stub
    assetsymbol="VAL"
    isformatted="true"
    label="Minimum Received"
    labeltooltip="Your transaction will revert if there is a large, unfavorable price movement before it is confirmed."
    value="1.2345678"
  />
    
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="Price Impact"
    labeltooltip="The difference between the market price and estimated price due to trade size."
    value=""
  >
    <div
      class="value-status-wrapper"
    >
      <formatted-amount-stub
        assetsymbol=""
        class="swap-value"
        fontsizerate="normal"
        fontweightrate="normal"
        value="-0.1"
      >
        %
      </formatted-amount-stub>
    </div>
  </info-line-stub>
   
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="Route"
    labeltooltip=""
    value=""
  />
   
  <info-line-stub
    assetsymbol=" XOR"
    fiatvalue=""
    isformatted="true"
    label="Liquidity Provider Fee"
    labeltooltip="A portion of each trade ({liquidityProviderFee}%) goes to liquidity providers as a protocol incentive."
    value="0.003"
  />
   
  <!---->
</div>
`;

exports[`SwapInfo.vue [With Rewards]: should be rendered correctly 1`] = `
<div
  class="swap-info-container"
>
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="{first} per {second}"
    labeltooltip=""
    value="0.4"
  />
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="{first} per {second}"
    labeltooltip=""
    value="2.2"
  />
   
  <info-line-stub
    assetsymbol="VAL"
    isformatted="true"
    label="Minimum Received"
    labeltooltip="Your transaction will revert if there is a large, unfavorable price movement before it is confirmed."
    value="1.2345678"
  />
   
  <info-line-stub
    assetsymbol="XOR"
    fiatvalue="1.23"
    label="PSWAP Strategic Rewards"
    labeltooltip=""
    value="1"
  />
   
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="Price Impact"
    labeltooltip="The difference between the market price and estimated price due to trade size."
    value=""
  >
    <div
      class="value-status-wrapper"
    >
      <formatted-amount-stub
        assetsymbol=""
        class="swap-value"
        fontsizerate="normal"
        fontweightrate="normal"
        value="-0.1"
      >
        %
      </formatted-amount-stub>
    </div>
  </info-line-stub>
   
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="Route"
    labeltooltip=""
    value=""
  />
   
  <info-line-stub
    assetsymbol=" XOR"
    fiatvalue=""
    isformatted="true"
    label="Liquidity Provider Fee"
    labeltooltip="A portion of each trade ({liquidityProviderFee}%) goes to liquidity providers as a protocol incentive."
    value="0.003"
  />
   
  <info-line-stub
    assetsymbol=" XOR"
    fiatvalue="0.000861"
    isformatted="true"
    label="Network Fee"
    labeltooltip="Network fee is used to ensure SORA system's growth and stable performance."
    value="0.0007"
  />
</div>
`;

exports[`SwapInfo.vue [With Warning priceImpact]: should be rendered correctly 1`] = `
<div
  class="swap-info-container"
>
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="{first} per {second}"
    labeltooltip=""
    value="0.4"
  />
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="{first} per {second}"
    labeltooltip=""
    value="2.2"
  />
   
  <info-line-stub
    assetsymbol="VAL"
    isformatted="true"
    label="Minimum Received"
    labeltooltip="Your transaction will revert if there is a large, unfavorable price movement before it is confirmed."
    value="1.2345678"
  />
    
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="Price Impact"
    labeltooltip="The difference between the market price and estimated price due to trade size."
    value=""
  >
    <div
      class="value-status-wrapper warning"
    >
      <formatted-amount-stub
        assetsymbol=""
        class="swap-value"
        fontsizerate="normal"
        fontweightrate="normal"
        value="-1.8"
      >
        %
      </formatted-amount-stub>
    </div>
  </info-line-stub>
   
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="Route"
    labeltooltip=""
    value=""
  />
   
  <info-line-stub
    assetsymbol=" XOR"
    fiatvalue=""
    isformatted="true"
    label="Liquidity Provider Fee"
    labeltooltip="A portion of each trade ({liquidityProviderFee}%) goes to liquidity providers as a protocol incentive."
    value="0.003"
  />
   
  <info-line-stub
    assetsymbol=" XOR"
    fiatvalue="0.000861"
    isformatted="true"
    label="Network Fee"
    labeltooltip="Network fee is used to ensure SORA system's growth and stable performance."
    value="0.0007"
  />
</div>
`;

exports[`SwapInfo.vue [With isExchangeB]: should be rendered correctly 1`] = `
<div
  class="swap-info-container"
>
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="{first} per {second}"
    labeltooltip=""
    value="0.4"
  />
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="{first} per {second}"
    labeltooltip=""
    value="2.2"
  />
   
  <info-line-stub
    assetsymbol="XOR"
    fiatvalue="1.5185185"
    isformatted="true"
    label="Maximum Sold"
    labeltooltip="Your transaction will revert if there is a large, unfavorable price movement before it is confirmed."
    value="1.2345678"
  />
    
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="Price Impact"
    labeltooltip="The difference between the market price and estimated price due to trade size."
    value=""
  >
    <div
      class="value-status-wrapper"
    >
      <formatted-amount-stub
        assetsymbol=""
        class="swap-value"
        fontsizerate="normal"
        fontweightrate="normal"
        value="-0.1"
      >
        %
      </formatted-amount-stub>
    </div>
  </info-line-stub>
   
  <info-line-stub
    assetsymbol=""
    fiatvalue=""
    label="Route"
    labeltooltip=""
    value=""
  />
   
  <info-line-stub
    assetsymbol=" XOR"
    fiatvalue=""
    isformatted="true"
    label="Liquidity Provider Fee"
    labeltooltip="A portion of each trade ({liquidityProviderFee}%) goes to liquidity providers as a protocol incentive."
    value="0.003"
  />
   
  <info-line-stub
    assetsymbol=" XOR"
    fiatvalue="0.000861"
    isformatted="true"
    label="Network Fee"
    labeltooltip="Network fee is used to ensure SORA system's growth and stable performance."
    value="0.0007"
  />
</div>
`;
